### PR TITLE
[bigquery] Implement Storage Write API

### PR DESCRIPTION
**Pull request by nathan-artie**

### CODE REVIEW

File: clients/bigquery/bigquery.go [hunk 1]
```diff
@@ -163,7 +166,63 @@ func (s *Store) putTableViaLegacyAPI(ctx context.Context, tableID TableIdentifie
...
+	if err != nil {
+		return fmt.Errorf("failed to create managed stream")
+	}
+	defer managedStream.Close()
+
+	// TODO: Think about whether we want to batch the rows here.
+	// The size of a single [AppendRows] request must be lest than 10MB.
+	rows := tableData.Rows()
+	encoded := make([][]byte, len(rows))
+	for i, row := range rows {
```
Comment (Tang8330) [inline clients/bigquery/bigquery.go:203]: Can we chunk this?

File: clients/bigquery/bigquery.go [hunk 2]
```diff
@@ -163,7 +166,63 @@ func (s *Store) putTableViaLegacyAPI(ctx context.Context, tableID TableIdentifie
...
+	if err != nil {
+		return fmt.Errorf("failed to create managedwriter client: %w", err)
+	}
+	defer managedWriterClient.Close()
+
+	managedStream, err := managedWriterClient.NewManagedStream(ctx,
+		managedwriter.WithDestinationTable(
+			managedwriter.TableParentFromParts(bqTableID.ProjectID(), bqTableID.Dataset(), bqTableID.Table()),
+		),
+		managedwriter.WithType(managedwriter.DefaultStream), // TODO: Look into using other stream types
```
Comment (Tang8330) [inline clients/bigquery/bigquery.go:190]: DefaultStream should be correct https://cloud.google.com/bigquery/docs/write-api#selecting_a_type

File: clients/bigquery/bigquery.go [hunk 3]
```diff
@@ -163,7 +166,63 @@ func (s *Store) putTableViaLegacyAPI(ctx context.Context, tableID TableIdentifie
...
+		managedwriter.WithSchemaDescriptor(schemaDescriptor),
+		managedwriter.EnableWriteRetries(true),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create managed stream")
```
Comment (Tang8330) [inline clients/bigquery/bigquery.go:195]: Didn't include error

File: clients/bigquery/bigquery.go [hunk 4]
```diff
@@ -163,7 +166,63 @@ func (s *Store) putTableViaLegacyAPI(ctx context.Context, tableID TableIdentifie
...
+	defer managedStream.Close()
+
+	// TODO: Think about whether we want to batch the rows here.
+	// The size of a single [AppendRows] request must be lest than 10MB.
+	rows := tableData.Rows()
+	encoded := make([][]byte, len(rows))
+	for i, row := range rows {
+		message, err := rowToMessage(row, columns, *messageDescriptor, s.AdditionalDateFormats())
+		if err != nil {
+			return err
```
Comment (Tang8330) [inline clients/bigquery/bigquery.go:206]: Wrap the error?

File: clients/bigquery/storagewrite.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+	switch column.KindDetails.Kind {
+	case typing.Boolean.Kind:
+		fieldType = storagepb.TableFieldSchema_BOOL
+	case typing.Integer.Kind:
+		fieldType = storagepb.TableFieldSchema_INT64
+	case typing.Float.Kind:
+		fieldType = storagepb.TableFieldSchema_DOUBLE
+	case typing.EDecimal.Kind:
+		fieldType = storagepb.TableFieldSchema_STRING
```
Comment (Tang8330) [inline clients/bigquery/storagewrite.go:31]: What does this mean?

File: clients/bigquery/storagewrite.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+			fieldType = storagepb.TableFieldSchema_TIME
+		case ext.DateKindType:
+			fieldType = storagepb.TableFieldSchema_DATE
+		case ext.DateTimeKindType:
+			fieldType = storagepb.TableFieldSchema_TIMESTAMP
+		default:
+			return nil, fmt.Errorf("unsupported extended time details type: %q", column.KindDetails.ExtendedTimeDetails.Type)
+		}
+	case typing.Struct.Kind:
+		fieldType = storagepb.TableFieldSchema_STRING
```
Comment (Tang8330) [inline clients/bigquery/storagewrite.go:46]: Struct is also a string?

File: clients/bigquery/storagewrite.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+	if !ok {
+		return nil, fmt.Errorf("proto descriptor is not a message descriptor")
+	}
+	return &messageDescriptor, nil
+}
+
+// From https://cloud.google.com/java/docs/reference/google-cloud-bigquerystorage/latest/com.google.cloud.bigquery.storage.v1.CivilTimeEncoder
+// And https://cloud.google.com/pubsub/docs/bigquery#date_time_int
+func encodePacked64TimeMicros(value time.Time) int64 {
+	var result = int64(value.Nanosecond() / 1000)
```
Comment (Tang8330) [inline clients/bigquery/storagewrite.go:86]: `result := int64(...)`

Also can you leave a comment about what you're doing here?

File: clients/bigquery/storagewrite.go (hunk 5 above)
Comment (Tang8330) [reply clients/bigquery/storagewrite.go:31]: Based on the precision/scale it should be either NUMERIC(p, s) or string right?

File: clients/bigquery/storagewrite.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+			return nil, fmt.Errorf("failed to find a field named %q", column.Name())
+		}
+
+		value := row[column.Name()]
+
+		if value == nil {
+			continue
+		}
+
+		switch column.KindDetails.Kind {
```
Comment (Tang8330) [inline clients/bigquery/storagewrite.go:108]: Can we pull this out into a separate function?

Comment (Tang8330) [reply clients/bigquery/storagewrite.go:108]: And have it return the value so the caller can run `message.Set(field, value)`?

File: clients/bigquery/storagewrite.go [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+				message.Set(field, protoreflect.ValueOfInt64(encodePacked64TimeMicros(extTime.Time)))
+			case ext.DateKindType:
+				daysSinceEpoch := extTime.Unix() / (60 * 60 * 24)
+				message.Set(field, protoreflect.ValueOfInt32(int32(daysSinceEpoch)))
+			case ext.DateTimeKindType:
+				ts := timestamppb.New(extTime.Time)
+				if err := ts.CheckValid(); err != nil {
+					return nil, err
+				}
+				message.Set(field, protoreflect.ValueOfInt64(extTime.UnixMicro()))
```
Comment (nathan-artie) [inline clients/bigquery/storagewrite.go:173]: Huh this is weird, I'm not using `ts`.

File: clients/bigquery/storagewrite.go (hunk 8 above)
Comment (nathan-artie) [reply clients/bigquery/storagewrite.go:108]: I had that originally, and would have preferred to keep it that way, but the `typing.Array` type needs to be able to access the entire message: https://github.com/artie-labs/transfer/pull/715/files#diff-1340c750bdd06d343441e36dbe9c4877ca8fe7c9d09b2564227e40fe883f5710R190

File: clients/bigquery/storagewrite.go (hunk 7 above)
Comment (nathan-artie) [reply clients/bigquery/storagewrite.go:86]: Oops I thought I had included a URL that linked to the code I was referencing, let me see if I can find it.

File: clients/bigquery/storagewrite.go (hunk 8 above)
Comment (Tang8330) [reply clients/bigquery/storagewrite.go:108]: Got it

**Review by Tang8330 — APPROVED** [review]

File: clients/bigquery/storagewrite.go (hunk 5 above)
Comment (nathan-artie) [reply clients/bigquery/storagewrite.go:33]: This is just the type that we are using to transfer the data, it doesn't have to directly correspond to the type of the data in the database.

File: clients/bigquery/storagewrite.go (hunk 6 above)
Comment (nathan-artie) [reply clients/bigquery/storagewrite.go:48]: Yeah we get errors if we try to use `storagepb.TableFieldSchema_JSON` here.

**Pull request merged**